File: src/main/java/cube8540/oauth/authentication/credentials/oauth/security/endpoint/AuthorizationImplicitResponseEnhancer.java
package cube8540.oauth.authentication.credentials.oauth.security.endpoint;

import cube8540.oauth.authentication.credentials.oauth.security.AuthorizationRequest;
import cube8540.oauth.authentication.credentials.oauth.security.OAuth2AccessTokenDetails;
import cube8540.oauth.authentication.credentials.oauth.security.OAuth2AccessTokenGranter;
import cube8540.oauth.authentication.credentials.oauth.security.OAuth2ClientDetails;
import cube8540.oauth.authentication.credentials.oauth.security.OAuth2ClientDetailsService;
import cube8540.oauth.authentication.credentials.oauth.security.OAuth2TokenRequest;
import org.springframework.security.oauth2.core.AuthorizationGrantType;
import org.springframework.security.oauth2.core.endpoint.OAuth2AuthorizationResponseType;
import org.springframework.web.servlet.ModelAndView;
import org.springframework.web.servlet.view.RedirectView;

import java.net.URI;
import java.util.Set;

public class AuthorizationImplicitResponseEnhancer implements AuthorizationResponseEnhancer {

    private OAuth2AccessTokenGranter tokenGranter;
    private OAuth2ClientDetailsService clientDetailsService;

    private AuthorizationResponseEnhancer nextEnhancer;

    public AuthorizationImplicitResponseEnhancer(OAuth2AccessTokenGranter tokenGranter, OAuth2ClientDetailsService clientDetailsService) {
        this.tokenGranter = tokenGranter;
        this.clientDetailsService = clientDetailsService;
    }

    @Override
    public AuthorizationResponseEnhancer setNext(AuthorizationResponseEnhancer handler) {
        this.nextEnhancer = handler;
        return this.nextEnhancer;
    }

    @Override
    public ModelAndView enhance(ModelAndView modelAndView, AuthorizationRequest request) {
        if (request.getResponseType().equals(OAuth2AuthorizationResponseType.TOKEN)) {
            OAuth2ClientDetails clientDetails = clientDetailsService.loadClientDetailsByClientId(request.getClientId());
            OAuth2AccessTokenDetails token = tokenGranter.grant(clientDetails, new ImplicitTokenRequest(request));
            enhanceRedirectUrl(modelAndView, request, token);
        }

        return nextEnhancer != null ? nextEnhancer.enhance(modelAndView, request) : modelAndView;
    }

    private void enhanceRedirectUrl(ModelAndView modelAndView, AuthorizationRequest authorizationRequest, OAuth2AccessTokenDetails token) {
        RedirectView view = (RedirectView) modelAndView.getView();
        String redirectUrl = view.getUrl() + "#access_token=" + token.getTokenValue() + "&token_type=" + token.getTokenType()
                + "&expires_in=" + token.getExpiresIn() + "&scope=" + String.join(" ", token.getScopes());
        if (authorizationRequest.getState() != null) {
            redirectUrl += "&state=" + authorizationRequest.getState();
        }
        view.setUrl(redirectUrl);
    }

    private static class ImplicitTokenRequest implements OAuth2TokenRequest {

        private final AuthorizationRequest request;

        private ImplicitTokenRequest(AuthorizationRequest request) {
            this.request = request;
        }

        @Override
        public AuthorizationGrantType getGrantType() {
            return AuthorizationGrantType.IMPLICIT;
        }

        @Override
        public String getUsername() {
            return request.getUsername();
        }

        @Override
        public String getPassword() {
            return null;
        }

        @Override
        public String getClientId() {
            return request.getClientId();
        }

        @Override
        public String getRefreshToken() {
            return null;
        }

        @Override
        public String getCode() {
            return null;
        }

        @Override
        public String getState() {
            return request.getState();
        }

        @Override
        public URI getRedirectUri() {
            return request.getRedirectUri();
        }

        @Override
        public Set<String> getScopes() {
            return request.getRequestScopes();
        }
    }
}

File: src/main/kotlin/cube8540/oauth/authentication/credentials/oauth/security/endpoint/AuthorizationImplicitResponseEnhancer.kt
package cube8540.oauth.authentication.credentials.oauth.security.endpoint

import cube8540.oauth.authentication.credentials.oauth.security.AuthorizationRequest
import cube8540.oauth.authentication.credentials.oauth.security.OAuth2AccessTokenDetails
import cube8540.oauth.authentication.credentials.oauth.security.OAuth2AccessTokenGranter
import cube8540.oauth.authentication.credentials.oauth.security.OAuth2ClientDetailsService
import cube8540.oauth.authentication.credentials.oauth.security.OAuth2TokenRequest
import org.springframework.security.oauth2.core.AuthorizationGrantType
import org.springframework.security.oauth2.core.endpoint.OAuth2AuthorizationResponseType
import org.springframework.web.servlet.ModelAndView
import org.springframework.web.servlet.view.RedirectView
import java.net.URI

class AuthorizationImplicitResponseEnhancer(
    private val tokenGranter: OAuth2AccessTokenGranter,
    private val clientDetailsService: OAuth2ClientDetailsService,
    var nextEnhancer: AuthorizationResponseEnhancer?
): AuthorizationResponseEnhancer {

    constructor(tokenGranter: OAuth2AccessTokenGranter, clientDetailsService: OAuth2ClientDetailsService):
            this(tokenGranter, clientDetailsService, null)

    override fun setNext(handler: AuthorizationResponseEnhancer): AuthorizationResponseEnhancer =
        handler.also { this.nextEnhancer = handler }

    override fun enhance(modelAndView: ModelAndView, request: AuthorizationRequest): ModelAndView {
        if (request.responseType?.equals(OAuth2AuthorizationResponseType.TOKEN) == true) {
            val clientDetails = clientDetailsService.loadClientDetailsByClientId(request.clientId!!)
            val token = tokenGranter.grant(clientDetails, ImplicitTokenRequest(request))
            enhanceRedirectUrl(modelAndView, request, token)
        }

        return nextEnhancer?.enhance(modelAndView, request) ?: modelAndView
    }

    private fun enhanceRedirectUrl(modelAndView: ModelAndView, authorizationRequest: AuthorizationRequest, token: OAuth2AccessTokenDetails) {
        val view = modelAndView.view as RedirectView
        var redirectUrl = (view.url + "#access_token=" + token.tokenValue + "&token_type=" + token.tokenType
                + "&expires_in=" + token.expiresIn + "&scope=" + token.scopes?.joinToString(" "))
        if (authorizationRequest.state != null) {
            redirectUrl += "&state=" + authorizationRequest.state
        }
        view.url = redirectUrl
    }

    private inner class ImplicitTokenRequest(request: AuthorizationRequest): OAuth2TokenRequest {
        override val grantType: AuthorizationGrantType = AuthorizationGrantType.IMPLICIT
        override val username: String? = request.username
        override val password: String? = null
        override val clientId: String? = request.clientId
        override val refreshToken: String? = null
        override val code: String? = null
        override val state: String? = request.state
        override val redirectUri: URI? = request.redirectUri
        override val scopes: Set<String>? = request.requestScopes
    }
}